databricks_cli/dbfs/api.py

```diff
@@ -75,12 +88,40 @@ class DbfsErrorCodes(object):
     RESOURCE_DOES_NOT_EXIST = 'RESOURCE_DOES_NOT_EXIST'
     RESOURCE_ALREADY_EXISTS = 'RESOURCE_ALREADY_EXISTS'
     PARTIAL_DELETE = 'PARTIAL_DELETE'
+    TOO_MANY_REQUESTS = 'TOO_MANY_REQUESTS'
+
+
+def my_before_sleep(retry_state):
+    if retry_state.attempt_number < 1:
+        loglevel = logging.INFO
+    else:
+        loglevel = logging.WARNING
+    logger.log(
+        loglevel, ' Received 429 REQUEST_LIMIT_EXCEEDED for attempt %s. Retrying in %s seconds.',
+        retry_state.attempt_number, retry_state.idle_for)
```

Comment:
It might be that `retry_state.idle_for` represents the time we already slept before the current attempt and not the time before the next one. It's not clear from the docs. Did you check this by testing?

Reply:
Good point. The docs are indeed not clear. Note, There is no `after_sleep` function. Also, the docs say that the `retry_state` parameter includes information about the current invocation. So it makes sense that the idle_for would refer to how much it would be idle for before sleeping. So I think this is right, although I didn't test it manually. 
```
The callback should accept one parameter called retry_state that contains all information about current retry invocation.
```

Reply:
Fixed, we had use our own implementation. No out of the box parameters were available. See https://github.com/databricks/databricks-cli/pull/319/commits/f32120278157c6911938e056900c3267dd44ff7d#diff-dad49ab19367f9bdc72941b5d6b02c2fR103

Reply:
Thanks for addressing this comment. I validated the timeouts are correct by printing the time before and after running the function.
```
2020-09-22 15:40:54.885647 Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: 2020-09-22 15:40:54.885906 Received 429 REQUEST_LIMIT_EXCEEDED for attempt 1. Retrying in 0.550553169504 seconds.
2020-09-22 15:40:55.437166 Executing func.
2020-09-22 15:40:55.479119 Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: 2020-09-22 15:40:55.479274 Received 429 REQUEST_LIMIT_EXCEEDED for attempt 2. Retrying in 0.367841898264 seconds.
2020-09-22 15:40:55.847585 Executing func.
2020-09-22 15:40:55.895012 Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: 2020-09-22 15:40:55.895145 Received 429 REQUEST_LIMIT_EXCEEDED for attempt 3. Retrying in 0.366708191017 seconds.
2020-09-22 15:40:56.262310 Executing func.
2020-09-22 15:40:56.304895 Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: 2020-09-22 15:40:56.305025 Received 429 REQUEST_LIMIT_EXCEEDED for attempt 4. Retrying in 7.51642709122 seconds.
2020-09-22 15:41:03.829079 Executing func.
2020-09-22 15:41:03.898720 Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: 2020-09-22 15:41:03.898903 Received 429 REQUEST_LIMIT_EXCEEDED for attempt 5. Retrying in 10.6680429268 seconds.
2020-09-22 15:41:14.577751 Executing func.
2020-09-22 15:41:14.622335 Rate limit exceeded. Retrying with exponential backoff.
WARNING:databricks_cli.dbfs.api: 2020-09-22 15:41:14.622513 Received 429 REQUEST_LIMIT_EXCEEDED for attempt 6. Retrying in 28.4351091639 seconds.
```